lib/gc/ExecutionEngine/GPURuntime/ocl/GpuOclRuntime.cpp

```diff
@@ -0,0 +1,690 @@
+    err = clSetKernelExecInfo(kernel,
+                              CL_KERNEL_EXEC_INFO_INDIRECT_HOST_ACCESS_INTEL,
+                              sizeof(enable), &enable);
+    CL_CHECKR(err, "Failed to set indirect host access.");
+    err = clSetKernelExecInfo(kernel,
+                              CL_KERNEL_EXEC_INFO_INDIRECT_DEVICE_ACCESS_INTEL,
+                              sizeof(enable), &enable);
+    CL_CHECKR(err, "Failed to set indirect device access.");
+    err = clSetKernelExecInfo(kernel,
+                              CL_KERNEL_EXEC_INFO_INDIRECT_SHARED_ACCESS_INTEL,
+                              sizeof(enable), &enable);
+    CL_CHECKR(err, "Failed to set indirect shared access.");
+
+    return new Kernel(program, kernel, gridSize, blockSize, argNum, argSize);
+  }
+
+  static void kernelDestroy(size_t count, Kernel **kernels) {
+    gcLogD("Destroying kernels.");
+    for (size_t i = 0; i < count; i++) {
+      if (kernels[i]) {
+        delete kernels[i];
+      }
+    }
+  }
+
+  static void kernelLaunch(OclContext *ctx, Kernel *kernel, ...) {
+    struct ClonedKernel {
+      cl_kernel kernel;
+
+      explicit ClonedKernel(cl_kernel kernel) : kernel(kernel) {}
+
+      ~ClonedKernel() {
+        gcLogD("Releasing cloned OpenCL kernel: ", kernel);
+        clReleaseKernel(kernel);
+      }
+    };
+
+    gcLogD("Launching kernel: ", kernel->kernel);
+
+    cl_int err;
+    ClonedKernel cloned{clCloneKernel(kernel->kernel, &err)};
+    CL_CHECKR(err, "Failed to clone OpenCL kernel: ", kernel->kernel);
+    gcLogD("Cloned OpenCL kernel ", kernel->kernel, ": ", cloned.kernel);
+
+    va_list args;
+    va_start(args, kernel);
+    for (size_t i = 0, n = kernel->argSize.size(); i < n; i++) {
+      auto size = kernel->argSize[i];
+      switch (size) {
+      case 0: {
+        void *ptr = va_arg(args, void *);
+        if (ctx->clPtrs.find(ptr) == ctx->clPtrs.end()) {
+          gcLogD("Setting kernel ", cloned.kernel, " argument ", i,
+                 " to USM pointer ", ptr);
+          err = ctx->runtime.ext.clSetKernelArgMemPointerINTEL(cloned.kernel, i,
+                                                               ptr);
+        } else {
+          gcLogD("Setting kernel ", cloned.kernel, " argument ", i,
+                 " to CL pointer ", ptr);
+          err = clSetKernelArg(cloned.kernel, i, sizeof(cl_mem), &ptr);
+        }
+        break;
+      }
+      case 1: {
+        auto val = static_cast<int8_t>(va_arg(args, int));
+        gcLogD("Setting kernel ", cloned.kernel, " argument ", i, " to ", val);
+        err = clSetKernelArg(cloned.kernel, i, size, &val);
+        break;
+      }
+      case 2: {
+        auto val = static_cast<int8_t>(va_arg(args, int));
+        gcLogD("Setting kernel ", cloned.kernel, " argument ", i, " to ", val);
+        err = clSetKernelArg(cloned.kernel, i, size, &val);
+        break;
+      }
+      case 4: {
+        auto val = va_arg(args, int32_t);
```

Comment:
I assume that `int32_t` is compatible with `float` right and `int64_t` is compatible with `double` in our case, right?

[cpp reference:](https://en.cppreference.com/w/cpp/utility/variadic/va_arg#:~:text=If%20the%20type)
> If the type of the next argument in ap (after promotions) is not compatible with T, the behavior is undefined

Reply:
Now it's expected, that all the args are pointers. The actual values are put on the stack in MLIR.